File: roxconnector_plugin/roxcomposer_message.js
let proto = require('./service_com_pb.js');
let uuid = require('uuid/v4');

exports = module.exports = {};

exports.Service = Service;
exports.Message = Message;
exports.deserialize_from_json = deserialize_from_json;
exports.deserialize_from_protobuf = deserialize_from_protobuf;
exports.deserialize = deserialize;
exports.get_packet_len = get_packet_len;

function get_packet_len(msg) {
	if(msg.length >= 4)
		return Buffer.from(msg).readUInt32BE(0);
	throw('argument too short');
}

function frame_message(msg) {
	let len = msg.length;
	let buf = Buffer.allocUnsafe(4 + len);
	buf.writeUInt32BE(len, 0);
	Buffer.from(msg.buffer).copy(buf, 4);

	return buf;
}

function unframe_message(msg) {
	let l = get_packet_len(msg);
    	if (l + 4 != msg.length)
        	throw (`Message of length ${msg.length} does not match frame header ${l}`);

	return new Uint8Array(msg.slice(4));
}

// ##### SERVICE CLASS CODE ######
function Service(ip, port, parameters=[]) {
	this.ip = ip;
	this.port = port;
	this.parameters = parameters;

	this.encodeId = encodeId;
	this.decodeId = decodeId;
}

function encodeId() {
	return `${this.ip}:${this.port}`;
}

// STATIC
function decodeId(idstring) {
	let parts = idstring.split(':');
	if (parts.length < 2) 
		throw `invalid service id: ${idstring}`;

	let port = parts.pop();
	let ip = parts.join(':');

	return { ip: ip, port: port };
}

// #### MESSAGE CLASS CODE ####
function Message() {
	this.pipeline = [];
	this.id = null;
	this.created = null;
	this.payload = null;

	this.create_id = create_id;
	this.set_payload = set_payload;
	this.get_payload = get_payload;
	this.get_content_as_dict = get_content_as_dict;
	this.pop_service = pop_service;
	this.peek_service = peek_service;
	this.add_service = add_service;
	this.has_empty_pipeline = has_empty_pipeline;
	this.serialize_to_protobuf = serialize_to_protobuf;
	this.serialize_to_json = serialize_to_json;
	this.serialize = serialize;

	this.create_id();
	this.created = Date.now();
}

function create_id() {
	this.id = uuid();
}

function set_payload(data) {
	this.payload = data
}

function get_payload() {
	return this.payload
}

function get_content_as_dict() {
	return {
		id: this.id,
		payload: this.payload,
		pipeline: this.pipeline
	};
}

function pop_service() {
	return this.pipeline.shift();
}

// returns undef if pipeline is empty
function peek_service() {
	return this.pipeline[0];
}

// returns undef if pipeline is empty
function add_service(service) {
	this.pipeline.push(service);
}

function has_empty_pipeline() {
	return this.pipeline.length == 0;
}

function serialize_to_protobuf() {
	pmsg = new proto.ROXcomposerMessage();
	let services = []
	for (let i in this.pipeline) {
		p = this.pipeline[i];
		let s = new proto.Service();
		s.setId(p.encodeId());
		s.setParametersList(p.parameters);
		services.push(s);
	}

	pmsg.setId(this.id);
	let payload = new proto.Payload();
	payload.setBody(this.payload);
	pmsg.setPayload(payload);
	let pipeline = new proto.Pipeline();
	pipeline.setServicesList(services);
	pmsg.setPipeline(pipeline);
	pmsg.setCreated(this.created);

	return pmsg.serializeBinary();
}

// STATIC
function deserialize_from_protobuf(binmsg) {
	let msg = new Message();
	let pmsg = proto.ROXcomposerMessage.deserializeBinary(binmsg);
	msg.set_payload(pmsg.getPayload().getBody());
	msg.id = pmsg.getId()
	let p = pmsg.getPipeline();
	let pipe = [];

	if (p) {
		pipe = p.getServicesList();
	}

	for (let s in pipe) {
		let addr = decodeId(pipe[s].getId());
		msg.add_service(new Service(addr.ip, addr.port, pipe[s].getParametersList()));
	}

	return msg;
}

function serialize_to_json() {
        let msg = {
			id: this.id,
			created: this.created,
			pipeline: this.pipeline.map( (x) => { return { id: x.encodeId(), parameters: x.parameters }; } ),
			payload: this.payload
            };

        return JSON.stringify(msg);
}

// STATIC
function deserialize_from_json(jsonstring) {
	let msg = new Message();
	let d = JSON.parse(jsonstring);
	msg.set_payload(d.payload);
	msg.id = d.id;
	msg.created = d.created;
	for (let i in d.pipeline) {
		let addr = decodeId(d.pipeline[i].id);
		msg.add_service(new Service(addr.ip, addr.port, d.pipeline[i].parameters));
	}

	return msg;
}

function serialize() {
	return frame_message(this.serialize_to_protobuf());
}

// STATIC
function deserialize(msg) {
    return this.deserialize_from_protobuf(unframe_message(msg));
}
        

File: roxcomposer/communication/roxcomposer_message.py
# |------------------- OPEN SOURCE LICENSE DISCLAIMER -------------------|
# |                                                                      |
# | Copyright (C) 2019  droxIT GmbH - devs@droxit.de                     |
# |                                                                      |
# | This file is part of ROXcomposer.                                    |
# |                                                                      |
# | ROXcomposer is free software: you can redistribute it and/or modify  |
# | it under the terms of the GNU Lesser General Public License as       |
# | published by the Free Software Foundation, either version 3 of the   |
# | License, or (at your option) any later version.                      |
# |                                                                      |
# | This program is distributed in the hope that it will be useful,      |
# | but WITHOUT ANY WARRANTY; without even the implied warranty of       |
# | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the         |
# | GNU General Public License for more details.                         |
# |                                                                      |
# | You have received a copy of the GNU Lesser General Public License    |
# | along with this program. See also <http://www.gnu.org/licenses/>.    |
# |                                                                      |
# |----------------------------------------------------------------------|

from roxcomposer.communication import service_com_pb2 as proto
import uuid
import json
import struct
import time
from roxcomposer import exceptions


def get_packet_len(msg):
    if len(msg) >= 4:
        return struct.unpack('>I', msg[:4])[0]
    raise exceptions.InvalidArgument('the provided string was too short: {}'.format(msg))


# prefixes msg with it's length as a 32bit big endian integer
def frame_message(msg):
    return struct.pack('>I', len(msg)) + msg


# remove the length prefix (see frame_message) from msg
# throws an exception if len(msg) does not match the prefixed length
def unframe_message(msg):
    l = struct.unpack('>I', msg[:4])[0]
    if l + 4 != len(msg):
        raise exceptions.MessageLengthMismatch(
            'Message of length {} does not match frame header {}'.format(len(msg), l))
    return msg[4:]


class Service:
    def __init__(self, ip, port, parameters=None):
        self.ip = ip
        self.port = port
        if parameters == None:
            self.parameters = []
        else:
            self.parameters = parameters

    def encodeId(self):
        return "{}:{}".format(self.ip, self.port)

    @staticmethod
    def decodeId(idstring):
        parts = idstring.split(':')
        if len(parts) < 2:
            errmsg = 'invalid service id: {}'.format(idstring)
            raise exceptions.InvalidServiceId(errmsg)
        port = int(parts.pop())
        ip = ':'.join(parts)
        return (ip, port)

    def __str__(self):
        return str({'ip': self.ip, 'port': self.port, 'parameters': self.parameters})

    def __repr__(self):
        return '<Service ' + str(self) + '>'

    def __eq__(self, other):
        return repr(self) == repr(other)


class Message:
    def __init__(self):
        self.pipeline = []
        self.id = None
        self.payload = None
        self.create_id()
        self.created = int(time.time() * 1000)

    def create_id(self):
        self.id = str(uuid.uuid4())

    def set_payload(self, data):
        self.payload = data

    def get_payload(self):
        return self.payload

    def get_content_as_dict(self):
        return {
            'id': self.id,
            'pipeline': self.pipeline,
            'payload': self.payload
        }

    # raises KeyError if the pipeline is empty
    def pop_service(self):
        return self.pipeline.pop(0)

    # raises KeyError if the pipeline is empty
    def peek_service(self):
        return self.pipeline[0]

    def add_service(self, service):
        self.pipeline.append(service)

    def has_empty_pipeline(self):
        return len(self.pipeline) == 0

    def serialize_to_protobuf(self):
        pmsg = proto.ROXcomposerMessage()
        for p in self.pipeline:
            s = pmsg.pipeline.services.add()
            s.id = p.encodeId()
            for para in p.parameters:
                param = s.parameters.add()
                param.serviceParams = para

        pmsg.id = self.id
        pmsg.created = self.created
        pmsg.payload.body = self.payload
        binmsg = pmsg.SerializeToString()
        return binmsg

    @staticmethod
    def deserialize_from_protobuf(binmsg):
        msg = Message()
        pmsg = proto.ROXcomposerMessage()
        pmsg.ParseFromString(binmsg)
        msg.set_payload(pmsg.payload.body)
        msg.id = pmsg.id
        msg.created = pmsg.created
        for s in pmsg.pipeline.services:
            ip, port = Service.decodeId(s.id)
            msg.add_service(Service(ip, port, [x.serviceParams for x in s.parameters]))

        return msg

    def serialize_to_json(self):
        msg = {
            'id': self.id,
            'created': self.created,
            'pipeline': [x for x in map(lambda s: {'id': s.encodeId(), 'parameters': s.parameters}, self.pipeline)],
            'payload': self.payload
        }

        return json.dumps(msg).encode()

    @staticmethod
    def deserialize_from_json(jsonstring):
        msg = Message()
        d = json.loads(jsonstring)
        msg.set_payload(d['payload'])
        msg.id = d['id']
        for s in d['pipeline']:
            ip, port = Service.decodeId(s['id'])
            msg.add_service(Service(ip, port, s['parameters']))

        return msg

    def serialize(self):
        return frame_message(self.serialize_to_protobuf())

    @staticmethod
    def deserialize(msg):
        return Message.deserialize_from_protobuf(unframe_message(msg))
